sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala

```diff
@@ -5,25 +5,146 @@ package com.azure.cosmos.implementation
+      .fromString(continuation)
+      .toJson()
+  }
+
+  def extractLsnFromChangeFeedContinuation(continuation: String): Long = {
     val lsnToken = ChangeFeedState
       .fromString(continuation)
       .getContinuation
       .getCurrentContinuationToken
       .getToken
 
+    toLsn(lsnToken)
+  }
+
+  def overrideLsnInChangeFeedContinuation(continuation: String, newContinuationToken: String): String = {
+    val state = ChangeFeedState
+      .fromString(continuation)
+    val continuationToken = state.getContinuation.getCurrentContinuationToken
+    continuationToken.setToken(newContinuationToken)
+    state.toString
+  }
+
+  def mergeChangeFeedContinuations(continuationTokens: Iterable[String]): String = {
+    var count = 0
+    val states = continuationTokens.map(s => {
+      count += 1
+      ChangeFeedState.fromString(s)
+    }).toArray
+
+    ChangeFeedState.merge(states).toString()
+  }
+
+  def createChangeFeedStateJson
+  (
+    startOffsetContinuationState: String,
+    feedRangeToLsn: Array[(NormalizedRange, Long)]
+  ): String = {
+
+    val continuationTokens = feedRangeToLsn
+      .map(rangeToLsn => new CompositeContinuationToken(
+        toContinuationToken(rangeToLsn._2),
+        toCosmosRange(rangeToLsn._1)))
+    val startState = ChangeFeedState.fromString(startOffsetContinuationState)
+    new ChangeFeedStateV1(
+      startState.getContainerRid,
+      startState.getFeedRange,
+      startState.getMode,
+      startState.getStartFromSettings,
+      FeedRangeContinuation.create(
+        startState.getContainerRid,
+        startState.getFeedRange,
+        continuationTokens.toList.asJava
+      )
+    ).toString
+  }
+
+  def toContinuationToken(lsn: Long): String = {
+    raw""""${String.valueOf(lsn)}""""
+  }
+
+  def extractContinuationTokensFromChangeFeedStateJson(stateJsonBase64: String): Array[(NormalizedRange, Long)] = {
+    assert(!Strings.isNullOrWhiteSpace(stateJsonBase64), s"Argument 'stateJsonBase64' must not be null or empty.")
+    val state = ChangeFeedState.fromString(stateJsonBase64)
+    state
+      .extractContinuationTokens() // already sorted
+      .asScala
+      .map(t => Tuple2(rangeToNormalizedRange(t.getRange), toLsn(t.getToken)))
+      .toArray
+  }
+
+  private[this] def rangeToNormalizedRange(rangeInput: Range[String]) = {
+    val range = FeedRangeInternal.normalizeRange(rangeInput)
+    assert(range != null, "Argument 'range' must not be null.")
+    assert(range.isMinInclusive, "Argument 'range' must be minInclusive")
+    assert(!range.isMaxInclusive, "Argument 'range' must be maxExclusive")
+
+    NormalizedRange(range.getMin, range.getMax)
+  }
+
+  def toLsn(lsnToken: String): Long = {
     // the continuation from the backend is encoded as '"<LSN>"' where LSN is a long integer
     // removing the first and last characters - which are the quotes
-    lsnToken.substring(1, lsnToken.length - 1).toLong
+    if (lsnToken != null && lsnToken.length > 2) {
```

Comment:
nit: I think in Scala using null is discouraged, rather use Options. Not critical though.

Reply:
True - and I used it in several places - but this one is a helper function to interop with java library - so Option would just be anotehr indriection.